### PR TITLE
Use ArrayAccessor in Comparison Kernels

### LINKED ISSUE #2135: Use ArrayAccessor in Comparison Kernels
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently the comparison kernels make extensive use of macros, with #2133 it is possible to rework these to instead make use of the `ArrayAccessor` trait. This should reduce code duplication and make the code easier to read and maintain.

**Describe the solution you'd like**

Rework the comparison kernels to use ArrayAccessor instead of `PrimitiveArray`, `GenericStringArray`, etc...

**Describe alternatives you've considered**

We could not do this

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #2135.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/compute/kernels/comparison.rs [hunk 1]
```diff
@@ -40,168 +39,72 @@ use regex::{escape, Regex};
...
-                null_bit_buffer,
-                0,
-                vec![Buffer::from(buffer)],
-                vec![],
-            )
-        };
-        Ok(BooleanArray::from(data))
-    }};
-}
+fn compare_op<T: ArrayAccessor, F>(left: T, right: T, op: F) -> Result<BooleanArray>
```
Comment (viirya) [inline arrow/src/compute/kernels/comparison.rs:44]: `ArrayAccessor` allows us to replace macro with generic function like this.

File: arrow/src/compute/kernels/comparison.rs [hunk 2]
```diff
@@ -1964,6 +1838,195 @@ macro_rules! typed_cmp {
...
+        (DataType::Boolean, DataType::Boolean) => compare_op(
+            left.as_any().downcast_ref::<BooleanArray>().unwrap(),
+            right.as_any().downcast_ref::<BooleanArray>().unwrap(),
+            |a, b| !(a ^ b),
+        ),
+        (DataType::Int8, DataType::Int8) => compare_op(
+            left.as_any().downcast_ref::<Int8Array>().unwrap(),
+            right.as_any().downcast_ref::<Int8Array>().unwrap(),
+            |a, b| a == b,
+        ),
```
Comment (viirya) [inline arrow/src/compute/kernels/comparison.rs:1852]: But for some macros which we match array types, seems it doesn't reduce too much duplication.

File: arrow/src/compute/kernels/comparison.rs [hunk 3]
```diff
@@ -1931,177 +1804,107 @@ where
     Ok(BooleanArray::from(data))
 }
 
-macro_rules! typed_cmp {
```
Comment (viirya) [inline arrow/src/compute/kernels/comparison.rs:1934]: We can remove this macro.

**Comment by viirya** [comment]
`cargo bench --bench comparison_kernels -- --baseline master` doesn't see regression.

**Comment by jhorstmann** [comment]
Nice improvement!

**Review by alamb — APPROVED** [review]
Looks very nice 

cc @tustvold

**Comment by viirya** [comment]
Thanks @jhorstmann @alamb @tustvold

**Pull request merged**